### PR TITLE
Add topology sort for both C++ and Python interface

### PR DESCRIPTION
**Pull request by fabratu**
This PR adds a sequential Topology Sort algorithm as requested/discussed here: https://github.com/networkit/networkit/discussions/829

### CODE REVIEW

File: networkit/cpp/graph/test/GraphToolsGTest.cpp [hunk 1]
```diff
@@ -1023,4 +1023,36 @@ TEST_P(GraphToolsGTest, testEdgesSortedByWeight) {
...
+    G.addEdge(0, 1);
+    G.addEdge(0, 2);
+    G.addEdge(2, 1);
+    G.addEdge(1, 3);
+    G.addEdge(4, 2);
+
+    if (directed()) {
+        std::vector<node> res = GraphTools::topologySort(G);
+        std::vector<node> test = {4, 0, 2, 1, 3};
+        EXPECT_TRUE(res == test);
```
Comment (opatut) [inline networkit/cpp/graph/test/GraphToolsGTest.cpp:1048]: This test looks rather unstable, as there are multiple possible results, and the actual returned order depends on the implementation, the actual choice from the possible results is undefined behaviour. 

The test could instead iterate through all edges *(u, v)* and assert that u's index in the result is smaller than v's index.

**Review by opatut — COMMENTED** [review]
Thanks for implementing this so quickly. Not sure what the requirements are for tests, but I see a case here for a little improvement for future test stability and correctness.

Otherwise, awesome work ;) (Haven't tested it yet though, no idea how to compile it yet).

File: networkit/cpp/graph/GraphTools.cpp [hunk 2]
```diff
@@ -504,5 +504,45 @@ void sortEdgesByWeight(Graph &G, bool decreasing) {
...
+    count current = G.numberOfNodes() - 1;
+    std::function<void(node)> exploreNode;
+
+    exploreNode = [&](node u) {
+        if (topSortMark[u] == nodeMark::PERM) {
+            return;
+        }
+        if (topSortMark[u] == nodeMark::TEMP) {
+            throw std::runtime_error(
+                "Error: the input graph has cycles. Therefore no valid topology can be computed.");
```
Comment (angriman) [inline networkit/cpp/graph/GraphTools.cpp:525]:
```suggestion
                "Error: the input graph has cycles.");
```

File: networkit/cpp/graph/GraphTools.cpp [hunk 3]
```diff
@@ -504,5 +504,45 @@ void sortEdgesByWeight(Graph &G, bool decreasing) {
...
+        }
+
+        topSortMark[u] = nodeMark::TEMP;
+
+        for (auto v : G.neighborRange(u)) {
+            exploreNode(v);
+        }
```
Comment (angriman) [inline networkit/cpp/graph/GraphTools.cpp:532]:
```suggestion
        G.forNeighbors(u, exploreNode);
```

File: include/networkit/graph/GraphTools.hpp [hunk 4]
```diff
@@ -395,7 +395,7 @@ std::unordered_map<node, node> getRandomContinuousNodeIds(const Graph &graph);
  * @param 	nodeIdMap	The mapping from node ids with gaps to continuous node ids (i.e. from
  * @getContinuousNodeIds)
  * @param 	G 			The compacted graph (currently only needed for the upper node id bound)
- * @return 				A vector of nodes id where the index is the node id of the compacted graph
+ * @return 				A vector of node-ids where the index is the node id of the compacted graph
```
Comment (angriman) [inline include/networkit/graph/GraphTools.hpp:398]: This change is not related to the PR.

File: include/networkit/graph/GraphTools.hpp [hunk 5]
```diff
@@ -420,6 +420,15 @@ Graph restoreGraph(const std::vector<node> &invertedIdMap, const Graph &G);
...
 void sortEdgesByWeight(Graph &G, bool decreasing = false);
 
+/**
+ * Given a directed graph G, the topology sort algorithm creates one valid topology order of nodes.
+ * Undirected graphs are not accepted as input, since a topology sort is a linear ordering of
+ * vertices such that for every edge u -> v, node u comes before v in the ordering.
+ * @param   G           Directed input graph
+ * @return              A vector of node-ids sorted according to their topology.
+ */
+std::vector<node> topologySort(const Graph &G);
```
Comment (angriman) [inline include/networkit/graph/GraphTools.hpp:430]:
```suggestion
std::vector<node> topologicalSort(const Graph &G);
```

File: networkit/graphtools.pyx [hunk 6]
```diff
@@ -43,6 +43,7 @@ cdef extern from "<networkit/graph/GraphTools.hpp>" namespace "NetworKit::GraphT
 	unordered_map[node,node] getContinuousNodeIds(_Graph G) nogil except +
 	unordered_map[node,node] getRandomContinuousNodeIds(_Graph G) nogil except +
 	void sortEdgesByWeight(_Graph G, bool_t) nogil except +
+	vector[node] topologySort(_Graph G) nogil except +
```
Comment (angriman) [inline networkit/graphtools.pyx:46]: Please add a simple Python test.

**Comment by avdgrinten** [comment]
In general, I would prefer to have this as an `Algorithm` subclass. This has the advantage that the state (= the `std::vector`s etc.) can be put into the class and re-used if multiple runs of the algorithm are required.

We can still have a helper function but it should internally just construct an object of the subclass and call `run()`.

File: networkit/cpp/graph/GraphTools.cpp [hunk 7]
```diff
@@ -504,5 +504,45 @@ void sortEdgesByWeight(Graph &G, bool decreasing) {
         });
 }
 
+std::vector<node> topologySort(const Graph &G) {
+    if (!G.isDirected()) {
+        throw std::runtime_error(
+            "Sorting according to topology is only defined for directed graphs.");
```
Comment (angriman) [inline networkit/cpp/graph/GraphTools.cpp:510]:
```suggestion
            "Topological sort is defined for directed graphs only.");
```

File: networkit/cpp/graph/GraphTools.cpp [hunk 8]
```diff
@@ -504,5 +504,43 @@ void sortEdgesByWeight(Graph &G, bool decreasing) {
...
+            "Topological sort is defined for directed graphs only.");
+    }
+
+    enum nodeMark { NONE, TEMP, PERM };
+    std::vector<nodeMark> topSortMark(G.numberOfNodes(), nodeMark::NONE);
+    std::vector<node> result(G.numberOfNodes(), 0);
+    count current = G.numberOfNodes() - 1;
+    std::function<void(node)> exploreNode;
+
+    exploreNode = [&](node u) {
```
Comment (avdgrinten) [inline networkit/cpp/graph/GraphTools.cpp:519]: This still uses `std::function` (which has a high overhead) even though that is not necessary:
```suggestion
    auto exploreNode = [&](node u) {
```

Comment (avdgrinten) [reply networkit/cpp/graph/GraphTools.cpp:519]: Ah wait, you're doing this to enable the recursion. It's probably better to use a normal function in this case (and not a lambda).

File: networkit/cpp/graph/TopologicalSort.cpp [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+void TopologicalSort::run() {
+    if (hasRun)
+        reset();
+
+    std::stack<node> nodeStack;
+
+    G->forNodes([&](node u) {
+        if (topSortMark[u] != NodeMark::PERM) {
```
Comment (angriman) [inline networkit/cpp/graph/TopologicalSort.cpp:28]: Just a minor thing here: to avoid indenting the whole function, I would write something like
```
if (topSortMark[u] == NodeMark::PERM)
    return;
// Rest of the code
```

File: networkit/cpp/graph/TopologicalSort.cpp [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+            nodeStack.push(u);
+            do {
+                node v = nodeStack.top();
+                if (topSortMark[v] != NodeMark::NONE) {
+                    nodeStack.pop();
+                    if (topSortMark[v] == NodeMark::TEMP) {
+                        topSortMark[v] = NodeMark::PERM;
+                        topology[current] = v;
+                        current--;
+                        assert(current >= 0);
```
Comment (angriman) [inline networkit/cpp/graph/TopologicalSort.cpp:38]: This assertion is not needed, `current` is an unsigned integer.

File: networkit/cpp/graph/test/TopologicalSortGTest.cpp [hunk 11]
```diff
@@ -0,0 +1,76 @@
+/*
+ * TopologicalSort.hpp
```
Comment (angriman) [inline networkit/cpp/graph/test/TopologicalSortGTest.cpp:2]: The file name is different

**Review by angriman — APPROVED** [review]
